zerver/views/auth.py

```diff
@@ -719,6 +743,20 @@ def login_page(
     # that disables the default behavior of redirecting logged-in users to the
     # logged-in app.
     is_preview = 'preview' in request.GET
+
+    extra_context = kwargs.pop('extra_context', {})
+    extra_context["next"] = next
+
+    action_key_from_post = request.POST.get('action_key')
+    if action_key_from_post:
+        # In case some error occurs in the form validation,
+        # user should still be able to retry.
+        extra_context['action_key'] = action_key_from_post
+
+    if extra_context.get("action_key") is not None:
+        # We want to authenticate user while letting the user to be logged in.
+        is_preview = True
+
```

Comment:
```suggestion
    extra_context = kwargs.pop('extra_context', {})
    extra_context["next"] = next

    action_key = request.POST.get('action_key')
    extra_context['action_key'] = action_key
    if action_key:
        # We want to authenticate user while letting the user to be logged in.
        is_preview = True
```

reads simpler and cleaner. Detailing that the key comes from POST in the variable name seems rather unnecessary too.

Reply:
The key only comes from POST only during email+password once the user clicks login button otherwise we send action key in through `extra_context` in the `start_email_change` view that's run on clicking confirmation link.

Doing that would replace the `action_key` with `None`.

Can we do something in the code to make this clear? 🤔

Reply:
Ah I see. Hmm well either way, the natural role `extra_context`  is to simply affect how the template will be rended. What you're doing with 
```
    if extra_context.get("action_key") is not None:
        # We want to authenticate user while letting the user to be logged in.
        is_preview = True
```

is altering the behavior of the login_page function with what's supposed to only affect the template. That's not a great way of conveying "don't redirect the user to home" to the function. One option is to not call `login_page` with extra_context in start_email_change, but rather redirect the user to `/login/` with `action_key` param in the `GET` and grab it from there.

Reply:
I like the idea of redirecting the user to `/login/` with `action_key` as a `GET` param and grabbing it from there. This also resolves the circular import error we had as we now do not want to import `login_page` from `auth.py`. So, I've removed that commit.